File: package/ios/React/CameraView.swift
//
//  CameraView.swift
//  mrousavy
//
//  Created by Marc Rousavy on 09.11.20.
//  Copyright © 2020 mrousavy. All rights reserved.
//

import AVFoundation
import Foundation
import UIKit
import MetalKit

// TODOs for the CameraView which are currently too hard to implement either because of AVFoundation's limitations, or my brain capacity
//
// CameraView+RecordVideo
// TODO: Better startRecording()/stopRecording() (promise + callback, wait for TurboModules/JSI)
//
// CameraView+TakePhoto
// TODO: Photo HDR

// MARK: - CameraView

public final class CameraView: UIView, CameraSessionDelegate, PreviewViewDelegate, FpsSampleCollectorDelegate {
    // pragma MARK: React Properties
    
    var lutTexture: MTLTexture?
    var lensLutTexture: MTLTexture?
    
    // props that require reconfiguring
    @objc var cameraId: NSString?
    @objc var enableDepthData = false
    @objc var enablePortraitEffectsMatteDelivery = false
    @objc var enableBufferCompression = false
    @objc var isMirrored = false
    @objc var lensLUTAsset: NSDictionary? {
        didSet {
            if lensLUTAsset == nil {
                lensLutTexture  = nil
                updatePreview()
                return
            }
            
            guard let lensLutUIImage = RCTConvert.uiImage(lensLUTAsset) else {
                return
            }
            
            guard let cgImage = lensLutUIImage.cgImage else {
                return
            }
            
            let textureLoader = MTKTextureLoader(device: MTLCreateSystemDefaultDevice()!)
            
            lensLutTexture = try? textureLoader.newTexture(cgImage: cgImage, options: [
                MTKTextureLoader.Option.SRGB : false
            ])
            
            if lensLutTexture == nil {
                print("Failed to create lens LUT texture")
            }
            else {
                print("Lens LUT Loaded")
            }
            
            updatePreview()
        }
    }
    @objc var lutAsset: NSDictionary? {
        didSet{
            // If lutAsset is nil, clear the LUT texture
            guard let lutAsset = lutAsset else {
                lutTexture = nil
                updatePreview()
                return
            }
            
            guard let lutUIImage = RCTConvert.uiImage(lutAsset) else {
                print("❌ Failed to convert LUT asset")
                lutTexture = nil
                updatePreview()
                return
            }
            
            guard let cgImage = lutUIImage.cgImage else {
                lutTexture = nil
                updatePreview()
                return
            }
            
            let textureLoader = MTKTextureLoader(device: MTLCreateSystemDefaultDevice()!)
            
            lutTexture = try? textureLoader.newTexture(cgImage: cgImage, options: [
                MTKTextureLoader.Option.SRGB : false
            ])
            if lutTexture == nil {
                print("Failed to create LUT texture")
            } else {
                print("LUT Loaded")
            }
            
            updatePreview()
        }
    }
    
    // use cases
    @objc var photo = false
    @objc var video = false
    @objc var audio = false
    @objc var enableFrameProcessor = false
    @objc var codeScannerOptions: NSDictionary?
    @objc var pixelFormat: NSString?
    @objc var enableLocation = false
    @objc var preview = true {
        didSet {
            updatePreview()
        }
    }
    
    @objc var fisheyeW = false
    @objc var fisheyeF = false
    
    // props that require format reconfiguring
    @objc var format: NSDictionary?
    @objc var minFps: NSNumber?
    @objc var maxFps: NSNumber?
    @objc var videoHdr = false
    @objc var photoHdr = false
    @objc var photoQualityBalance: NSString?
    @objc var lowLightBoost = false
    @objc var outputOrientation: NSString?
    @objc var videoBitRateOverride: NSNumber?
    @objc var videoBitRateMultiplier: NSNumber?
    
    // other props
    @objc var isActive = false
    @objc var torch = "off"
    @objc var zoom: NSNumber = 1.0 // in "factor"
    @objc var exposure: NSNumber = 0.0
    @objc var temperature: NSNumber = 5800.0
    @objc var tint: NSNumber = 0.0
    @objc var videoStabilizationMode: NSString?
    @objc var resizeMode: NSString = "cover" {
        didSet {
            updatePreview()
        }
    }
    
    // events
    @objc var onInitializedEvent: RCTDirectEventBlock?
    @objc var onErrorEvent: RCTDirectEventBlock?
    @objc var onStartedEvent: RCTDirectEventBlock?
    @objc var onStoppedEvent: RCTDirectEventBlock?
    @objc var onPreviewStartedEvent: RCTDirectEventBlock?
    @objc var onPreviewStoppedEvent: RCTDirectEventBlock?
    @objc var onShutterEvent: RCTDirectEventBlock?
    @objc var onPreviewOrientationChangedEvent: RCTDirectEventBlock?
    @objc var onOutputOrientationChangedEvent: RCTDirectEventBlock?
    @objc var onViewReadyEvent: RCTDirectEventBlock?
    @objc var onAverageFpsChangedEvent: RCTDirectEventBlock?
    @objc var onCodeScannedEvent: RCTDirectEventBlock?
    
    // zoom
    @objc var enableZoomGesture = false {
        didSet {
            if enableZoomGesture {
                addPinchGestureRecognizer()
            } else {
                removePinchGestureRecognizer()
            }
        }
    }
    
#if VISION_CAMERA_ENABLE_FRAME_PROCESSORS
    @objc public var frameProcessor: FrameProcessor?
#endif
    
    // pragma MARK: Internal Properties
    var cameraSession = CameraSession()
    var previewView: PreviewView?
    var previewMetalView: PreviewMetalView?
    var videoFilter: FilterRenderer?
    var isMounted = false
    private var currentConfigureCall: DispatchTime?
    private let fpsSampleCollector = FpsSampleCollector()
    
    // CameraView+Zoom
    var pinchGestureRecognizer: UIPinchGestureRecognizer?
    var pinchScaleOffset: CGFloat = 1.0
    
    // CameraView+TakeSnapshot
    var latestVideoFrame: Snapshot?
    
    // pragma MARK: Setup
    
    override public init(frame: CGRect) {
        super.init(frame: frame)
        cameraSession.delegate = self
        fpsSampleCollector.delegate = self
        updatePreview()
    }
    
    @available(*, unavailable)
    required init?(coder _: NSCoder) {
        fatalError("init(coder:) is not implemented.")
    }
    
    override public func willMove(toSuperview newSuperview: UIView?) {
        super.willMove(toSuperview: newSuperview)
        
        if newSuperview != nil {
            fpsSampleCollector.start()
            if !isMounted {
                isMounted = true
                onViewReadyEvent?(nil)
            }
        } else {
            fpsSampleCollector.stop()
        }
    }
    
    override public func layoutSubviews() {
        if let previewView {
            previewView.frame = frame
            previewView.bounds = bounds
        }
        if let previewMetalView {
            previewMetalView.frame = frame
            previewMetalView.bounds = bounds
        }
    }
    
    func getPixelFormat() -> PixelFormat {
        // TODO: Use ObjC RCT enum parser for this
        if let pixelFormat = pixelFormat as? String {
            do {
                return try PixelFormat(jsValue: pixelFormat)
            } catch {
                if let error = error as? CameraError {
                    onError(error)
                } else {
                    onError(.unknown(message: error.localizedDescription, cause: error as NSError))
                }
            }
        }
        return .yuv
    }
    
    func getTorch() -> Torch {
        // TODO: Use ObjC RCT enum parser for this
        if let torch = try? Torch(jsValue: torch) {
            return torch
        }
        return .off
    }
    
    func getPhotoQualityBalance() -> QualityBalance {
        if let photoQualityBalance = photoQualityBalance as? String,
           let balance = try? QualityBalance(jsValue: photoQualityBalance) {
            return balance
        }
        return .balanced
    }
    
    // pragma MARK: Props updating
    override public final func didSetProps(_ changedProps: [String]!) {
        VisionLogger.log(level: .info, message: "Updating \(changedProps.count) props: [\(changedProps.joined(separator: ", "))]")
        let now = DispatchTime.now()
        currentConfigureCall = now
        
        cameraSession.configure { [self] config in
            // Check if we're still the latest call to configure { ... }
            guard currentConfigureCall == now else {
                // configure waits for a lock, and if a new call to update() happens in the meantime we can drop this one.
                // this works similar to how React implemented concurrent rendering, the newer call to update() has higher priority.
                VisionLogger.log(level: .info, message: "A new configure { ... } call arrived, aborting this one...")
                throw CameraConfiguration.AbortThrow.abort
            }
            
            // Input Camera Device
            config.cameraId = cameraId as? String
            config.isMirrored = isMirrored
            
            // Photo
            if photo {
                config.photo = .enabled(config: CameraConfiguration.Photo(qualityBalance: getPhotoQualityBalance(),
                                                                          enableDepthData: enableDepthData,
                                                                          enablePortraitEffectsMatte: enablePortraitEffectsMatteDelivery))
            } else {
                config.photo = .disabled
            }
            
            // Video/Frame Processor
            if video || enableFrameProcessor {
                config.video = .enabled(config: CameraConfiguration.Video(pixelFormat: getPixelFormat(),
                                                                          enableBufferCompression: enableBufferCompression,
                                                                          enableHdr: videoHdr,
                                                                          enableFrameProcessor: enableFrameProcessor))
            } else {
                config.video = .disabled
            }
            
            // Audio
            if audio {
                config.audio = .enabled(config: CameraConfiguration.Audio())
            } else {
                config.audio = .disabled
            }
            
            // Code Scanner
            if let codeScannerOptions {
                let options = try CodeScannerOptions(fromJsValue: codeScannerOptions)
                config.codeScanner = .enabled(config: CameraConfiguration.CodeScanner(options: options))
            } else {
                config.codeScanner = .disabled
            }
            
            // Location tagging
            config.enableLocation = enableLocation && isActive
            
            // Video Stabilization
            if let jsVideoStabilizationMode = videoStabilizationMode as? String {
                let videoStabilizationMode = try VideoStabilizationMode(jsValue: jsVideoStabilizationMode)
                config.videoStabilizationMode = videoStabilizationMode
            } else {
                config.videoStabilizationMode = .off
            }
            
            // Orientation
            if let jsOrientation = outputOrientation as? String {
                let outputOrientation = try OutputOrientation(jsValue: jsOrientation)
                config.outputOrientation = outputOrientation
            } else {
                config.outputOrientation = .device
            }
            
            // Format
            if let jsFormat = format {
                let format = try CameraDeviceFormat(jsValue: jsFormat)
                config.format = format
            } else {
                config.format = nil
            }
            
            // Side-Props
            config.minFps = minFps?.int32Value
            config.maxFps = maxFps?.int32Value
            config.enableLowLightBoost = lowLightBoost
            config.torch = try Torch(jsValue: torch)
            
            // Zoom
            config.zoom = zoom.doubleValue
            
            // Exposure
            config.exposure = exposure.floatValue
            
            // Temperature
            config.temperature = temperature.floatValue
            
            // Tint
            config.tint = tint.floatValue
            
            // isActive
            config.isActive = isActive
        }
        
        // Store `zoom` offset for native pinch-gesture
        if changedProps.contains("zoom") {
            pinchScaleOffset = zoom.doubleValue
        }
        
        if let previewMetalView = previewMetalView {
            previewMetalView.fisheyeW = self.fisheyeW
        }
        
        if let previewMetalView = previewMetalView {
            previewMetalView.fisheyeF = self.fisheyeF
        }
        
        // Prevent phone from going to sleep
        UIApplication.shared.isIdleTimerDisabled = isActive
    }
    
    func updatePreview() {
        if preview && previewView == nil {
            // Always create PreviewView for normal camera preview
            previewView = cameraSession.createPreviewView(frame: frame)
            previewView!.delegate = self
            addSubview(previewView!)
            
            // Only add LUT filter and Metal view if LUT texture is available
            if let lutTexture = lutTexture {
                videoFilter = LutMetalRenderer(lutTexture: lutTexture,lensLutTexture: lensLutTexture!)
                previewMetalView = PreviewMetalView(frame: frame)
                addSubview(previewMetalView!)
                bringSubviewToFront(previewMetalView!)
                // Hide the normal preview when using LUT
                previewView!.isHidden = true
            } else {
                // No LUT, use normal preview
                videoFilter?.reset()
                videoFilter = nil
                previewMetalView?.removeFromSuperview()
                previewMetalView = nil
                previewView!.isHidden = false
            }
        } else if !preview && previewView != nil {
            // Remove PreviewView and destroy it
            previewView?.removeFromSuperview()
            previewView = nil
            previewMetalView?.removeFromSuperview()
            previewMetalView = nil
            videoFilter?.reset()
            videoFilter = nil
        } else if preview && previewView != nil {
            // Preview already exists, update LUT setup if needed
            if let lutTexture = lutTexture {
                // LUT is available, recreate filter with new texture (in case texture changed)
                videoFilter?.reset()
                videoFilter = LutMetalRenderer(lutTexture: lutTexture,lensLutTexture: lensLutTexture)
                if previewMetalView == nil {
                    previewMetalView = PreviewMetalView(frame: frame)
                    addSubview(previewMetalView!)
                    bringSubviewToFront(previewMetalView!)
                }
                // Hide the normal preview when using LUT
                previewView!.isHidden = true
            } else {
                // No LUT, remove filter and Metal view, show normal preview
                videoFilter?.reset()
                videoFilter = nil
                previewMetalView?.removeFromSuperview()
                previewMetalView = nil
                previewView!.isHidden = false
            }
        }
        
        if let previewView {
            // Update resizeMode from React
            let parsed = try? ResizeMode(jsValue: resizeMode as String)
            previewView.resizeMode = parsed ?? .cover
        }
    }
    
    
    // pragma MARK: Event Invokers
    
    func onError(_ error: CameraError) {
        VisionLogger.log(level: .error, message: "Invoking onError(): \(error.message)")
        
        var causeDictionary: [String: Any]?
        if case let .unknown(_, cause) = error,
           let cause = cause {
            causeDictionary = [
                "code": cause.code,
                "domain": cause.domain,
                "message": cause.description,
                "details": cause.userInfo,
            ]
        }
        onErrorEvent?([
            "code": error.code,
            "message": error.message,
            "cause": causeDictionary ?? NSNull(),
        ])
    }
    
    func onSessionInitialized() {
        onInitializedEvent?([:])
    }
    
    func onCameraStarted() {
        onStartedEvent?([:])
    }
    
    func onCameraStopped() {
        onStoppedEvent?([:])
    }
    
    func onPreviewStarted() {
        onPreviewStartedEvent?([:])
    }
    
    func onPreviewStopped() {
        onPreviewStoppedEvent?([:])
    }
    
    func onCaptureShutter(shutterType: ShutterType) {
        onShutterEvent?([
            "type": shutterType.jsValue,
        ])
    }
    
    func onOutputOrientationChanged(_ outputOrientation: Orientation) {
        onOutputOrientationChangedEvent?([
            "outputOrientation": outputOrientation.jsValue,
        ])
    }
    
    func onPreviewOrientationChanged(_ previewOrientation: Orientation) {
        onPreviewOrientationChangedEvent?([
            "previewOrientation": previewOrientation.jsValue,
        ])
    }
    
    func onFrame(sampleBuffer: CMSampleBuffer, orientation: Orientation, isMirrored: Bool) {
        // Update latest frame that can be used for snapshot capture
        latestVideoFrame = Snapshot(imageBuffer: sampleBuffer, orientation: orientation)
        
        // Notify FPS Collector that we just had a Frame
        fpsSampleCollector.onTick()
        
        
        
        // Only update Metal view if it exists (i.e., when using LUT)
        if let previewMetalView = previewMetalView {
            DispatchQueue.main.async {
                let interfaceOrientation = UIApplication.shared.connectedScenes.compactMap {$0 as? UIWindowScene}.first?.interfaceOrientation
                if let unwrappedVideoDataOutputConnection = self.cameraSession.videoOutput!.connection(with: .video) {
                    let videoDevicePosition = self.cameraSession.videoDeviceInput!.device.position
                    let rotation = PreviewMetalView.Rotation(with: interfaceOrientation!,
                                                             videoOrientation: unwrappedVideoDataOutputConnection.videoOrientation,
                                                             cameraPosition: videoDevicePosition)
                    previewMetalView.mirroring = (videoDevicePosition == .front)
                    if let rotation = rotation {
                        previewMetalView.rotation = rotation
                    }
                }
            }
            
            guard let videoPixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer),
                  let formatDescription = CMSampleBufferGetFormatDescription(sampleBuffer) else {
                return
            }
            
            var finalVideoPixelBuffer = videoPixelBuffer
            if let filter = videoFilter {
                if !filter.isPrepared {
                    /*
                     outputRetainedBufferCountHint is the number of pixel buffers the renderer retains. This value informs the renderer
                     how to size its buffer pool and how many pixel buffers to preallocate. Allow 3 frames of latency to cover the dispatch_async call.
                     */
                    filter.prepare(with: formatDescription, outputRetainedBufferCountHint: 3)
                }
                
                // Send the pixel buffer through the filter
                guard let filteredBuffer = filter.render(pixelBuffer: finalVideoPixelBuffer,fisheyeW: self.fisheyeW,fisheyeF: self.fisheyeF) else {
                    print("Unable to filter video buffer")
                    return
                }
                
                finalVideoPixelBuffer = filteredBuffer
            }
            
            previewMetalView.pixelBuffer = finalVideoPixelBuffer
        }
        
        
#if VISION_CAMERA_ENABLE_FRAME_PROCESSORS
        if let frameProcessor = frameProcessor {
            // Call Frame Processor
            let frame = Frame(buffer: sampleBuffer,
                              orientation: orientation.imageOrientation,
                              isMirrored: isMirrored)
            frameProcessor.call(frame)
        }
#endif
    }
    
    func onCodeScanned(codes: [CameraSession.Code], scannerFrame: CameraSession.CodeScannerFrame) {
        onCodeScannedEvent?([
            "codes": codes.map { $0.toJSValue() },
            "frame": scannerFrame.toJSValue(),
        ])
    }
    
    func onAverageFpsChanged(averageFps: Double) {
        onAverageFpsChangedEvent?([
            "averageFps": averageFps,
        ])
    }
}

File: package/android/src/main/java/com/mrousavy/camera/react/CameraView.kt
package com.mrousavy.camera.react

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import android.view.Gravity
import android.view.ScaleGestureDetector
import android.widget.FrameLayout
import androidx.camera.view.PreviewView
import com.google.mlkit.vision.barcode.common.Barcode
import com.mrousavy.camera.core.CameraConfiguration
import com.mrousavy.camera.core.CameraSession
import com.mrousavy.camera.core.CodeScannerFrame
import com.mrousavy.camera.core.types.CameraDeviceFormat
import com.mrousavy.camera.core.types.CodeScannerOptions
import com.mrousavy.camera.core.types.Orientation
import com.mrousavy.camera.core.types.OutputOrientation
import com.mrousavy.camera.core.types.PixelFormat
import com.mrousavy.camera.core.types.PreviewViewType
import com.mrousavy.camera.core.types.QualityBalance
import com.mrousavy.camera.core.types.ResizeMode
import com.mrousavy.camera.core.types.ShutterType
import com.mrousavy.camera.core.types.Torch
import com.mrousavy.camera.core.types.VideoStabilizationMode
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessor
import com.mrousavy.camera.react.extensions.installHierarchyFitter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

//
// TODOs for the CameraView which are currently too hard to implement either because of CameraX' limitations, or my brain capacity.
//
// TODO: High-speed video recordings (export in CameraViewModule::getAvailableVideoDevices(), and set in CameraView::configurePreview()) (120FPS+)
// TODO: Better startRecording()/stopRecording() (promise + callback, wait for TurboModules/JSI)
// TODO: takePhoto() depth data
// TODO: takePhoto() raw capture
// TODO: takePhoto() return with jsi::Value Image reference for faster capture
// TODO: Support videoCodec on Android

@SuppressLint("ClickableViewAccessibility", "ViewConstructor", "MissingPermission")
class CameraView(context: Context) :
  FrameLayout(context),
  CameraSession.Callback,
  FpsSampleCollector.Callback {
  companion object {
    const val TAG = "CameraView"
  }

  // react properties
  // props that require reconfiguring
  var cameraId: String? = null
  var enableDepthData = false
  var enablePortraitEffectsMatteDelivery = false
  var isMirrored = false

  // use-cases
  var photo = false
  var video = false
  var audio = false
  var enableFrameProcessor = false
  var pixelFormat: PixelFormat = PixelFormat.YUV
  var enableLocation = false
  var preview = true
    set(value) {
      field = value
      updatePreview()
    }

  // props that require format reconfiguring
  var format: CameraDeviceFormat? = null
  var minFps: Int? = null
  var maxFps: Int? = null
  var videoStabilizationMode: VideoStabilizationMode? = null
  var videoHdr = false
  var photoHdr = false
  var videoBitRateOverride: Double? = null
  var videoBitRateMultiplier: Double? = null

  // TODO: Use .BALANCED once CameraX fixes it https://issuetracker.google.com/issues/337214687
  var photoQualityBalance = QualityBalance.SPEED
  var lowLightBoost = false

  // other props
  var isActive = false
  var torch: Torch = Torch.OFF
  var zoom: Float = 1f // in "factor"
  var exposure: Double = 0.0
  var outputOrientation: OutputOrientation = OutputOrientation.DEVICE
  var androidPreviewViewType: PreviewViewType = PreviewViewType.SURFACE_VIEW
    set(value) {
      field = value
      updatePreview()
    }
  var enableZoomGesture = false
    set(value) {
      field = value
      updateZoomGesture()
    }
  var resizeMode: ResizeMode = ResizeMode.COVER
    set(value) {
      field = value
      updatePreview()
    }

  // code scanner
  var codeScannerOptions: CodeScannerOptions? = null

  // private properties
  private var isMounted = false
  private val mainCoroutineScope = CoroutineScope(Dispatchers.Main)

  // session
  internal val cameraSession: CameraSession
  internal var frameProcessor: FrameProcessor? = null
  internal var previewView: PreviewView? = null
  private var currentConfigureCall: Long = System.currentTimeMillis()
  private val fpsSampleCollector = FpsSampleCollector(this)

  init {
    clipToOutline = true
    cameraSession = CameraSession(context, this)
    this.installHierarchyFitter()
    updatePreview()
  }

  override fun onAttachedToWindow() {
    Log.i(TAG, "CameraView attached to window!")
    super.onAttachedToWindow()
    if (!isMounted) {
      // Notifies JS view that the native view is now available
      isMounted = true
      invokeOnViewReady()
    }
    // start collecting FPS samples
    fpsSampleCollector.start()
  }

  override fun onDetachedFromWindow() {
    Log.i(TAG, "CameraView detached from window!")
    super.onDetachedFromWindow()
    // stop collecting FPS samples
    fpsSampleCollector.stop()
  }

  fun destroy() {
    cameraSession.close()
  }

  fun update() {
    Log.i(TAG, "Updating CameraSession...")
    val now = System.currentTimeMillis()
    currentConfigureCall = now

    mainCoroutineScope.launch {
      cameraSession.configure { config ->
        if (currentConfigureCall != now) {
          // configure waits for a lock, and if a new call to update() happens in the meantime we can drop this one.
          // this works similar to how React implemented concurrent rendering, the newer call to update() has higher priority.
          Log.i(TAG, "A new configure { ... } call arrived, aborting this one...")
          throw CameraConfiguration.AbortThrow()
        }

        // Input Camera Device
        config.cameraId = cameraId

        // Preview
        val previewView = previewView
        if (previewView != null) {
          config.preview = CameraConfiguration.Output.Enabled.create(CameraConfiguration.Preview(previewView.surfaceProvider))
        } else {
          config.preview = CameraConfiguration.Output.Disabled.create()
        }

        // Photo
        if (photo) {
          config.photo = CameraConfiguration.Output.Enabled.create(CameraConfiguration.Photo(isMirrored, photoHdr, photoQualityBalance))
        } else {
          config.photo = CameraConfiguration.Output.Disabled.create()
        }

        // Video
        if (video || enableFrameProcessor) {
          config.video =
            CameraConfiguration.Output.Enabled.create(
              CameraConfiguration.Video(isMirrored, videoHdr, videoBitRateOverride, videoBitRateMultiplier)
            )
        } else {
          config.video = CameraConfiguration.Output.Disabled.create()
        }

        // Frame Processor
        if (enableFrameProcessor) {
          config.frameProcessor = CameraConfiguration.Output.Enabled.create(CameraConfiguration.FrameProcessor(isMirrored, pixelFormat))
        } else {
          config.frameProcessor = CameraConfiguration.Output.Disabled.create()
        }

        // Audio
        if (audio) {
          config.audio = CameraConfiguration.Output.Enabled.create(CameraConfiguration.Audio(Unit))
        } else {
          config.audio = CameraConfiguration.Output.Disabled.create()
        }

        // Location
        config.enableLocation = enableLocation && this@CameraView.isActive

        // Code Scanner
        val codeScanner = codeScannerOptions
        if (codeScanner != null) {
          config.codeScanner = CameraConfiguration.Output.Enabled.create(
            CameraConfiguration.CodeScanner(codeScanner.codeTypes)
          )
        } else {
          config.codeScanner = CameraConfiguration.Output.Disabled.create()
        }

        // Orientation
        config.outputOrientation = outputOrientation

        // Format
        config.format = format

        // Side-Props
        config.minFps = minFps
        config.maxFps = maxFps
        config.enableLowLightBoost = lowLightBoost
        config.torch = torch
        config.exposure = exposure

        // Zoom
        config.zoom = zoom

        // isActive
        config.isActive = this@CameraView.isActive
      }
    }
  }

  @SuppressLint("ClickableViewAccessibility")
  private fun updateZoomGesture() {
    if (enableZoomGesture) {
      val scaleGestureDetector = ScaleGestureDetector(
        context,
        object : ScaleGestureDetector.SimpleOnScaleGestureListener() {
          override fun onScale(detector: ScaleGestureDetector): Boolean {
            zoom *= detector.scaleFactor
            update()
            return true
          }
        }
      )
      setOnTouchListener { _, event ->
        scaleGestureDetector.onTouchEvent(event)
      }
    } else {
      setOnTouchListener(null)
    }
  }

  private fun updatePreview() {
    mainCoroutineScope.launch {
      if (preview && previewView == null) {
        // User enabled Preview, add the PreviewView
        previewView = createPreviewView()
        addView(previewView)
      } else if (!preview && previewView != null) {
        // User disabled Preview, remove the PreviewView
        removeView(previewView)
        previewView = null
      }
      previewView?.let {
        // Update implementation type from React
        it.implementationMode = androidPreviewViewType.toPreviewImplementationMode()
        // Update scale type from React
        it.scaleType = resizeMode.toScaleType()
      }
      update()
    }
  }

  private fun createPreviewView(): PreviewView =
    PreviewView(context).also {
      it.installHierarchyFitter()
      it.implementationMode = androidPreviewViewType.toPreviewImplementationMode()
      it.layoutParams = LayoutParams(
        LayoutParams.MATCH_PARENT,
        LayoutParams.MATCH_PARENT,
        Gravity.CENTER
      )
      var lastIsPreviewing = false
      it.previewStreamState.observe(cameraSession) { state ->
        Log.i(TAG, "PreviewView Stream State changed to $state")

        val isPreviewing = state == PreviewView.StreamState.STREAMING
        if (isPreviewing != lastIsPreviewing) {
          // Notify callback
          if (isPreviewing) {
            invokeOnPreviewStarted()
          } else {
            invokeOnPreviewStopped()
          }
          lastIsPreviewing = isPreviewing
        }
      }
    }

  override fun onFrame(frame: Frame) {
    // Update average FPS samples
    fpsSampleCollector.onTick()

    // Call JS Frame Processor
    frameProcessor?.call(frame)
  }

  override fun onError(error: Throwable) {
    invokeOnError(error)
  }

  override fun onInitialized() {
    invokeOnInitialized()
  }

  override fun onStarted() {
    invokeOnStarted()
  }

  override fun onStopped() {
    invokeOnStopped()
  }

  override fun onShutter(type: ShutterType) {
    invokeOnShutter(type)
  }

  override fun onOutputOrientationChanged(outputOrientation: Orientation) {
    invokeOnOutputOrientationChanged(outputOrientation)
  }

  override fun onPreviewOrientationChanged(previewOrientation: Orientation) {
    invokeOnPreviewOrientationChanged(previewOrientation)
  }

  override fun onCodeScanned(codes: List<Barcode>, scannerFrame: CodeScannerFrame) {
    invokeOnCodeScanned(codes, scannerFrame)
  }

  override fun onAverageFpsChanged(averageFps: Double) {
    invokeOnAverageFpsChanged(averageFps)
  }
}
